status-handler.js

```diff
@@ -143,6 +147,101 @@ class StatusHandler {
     this.log.info(`Ignored status event, no build occurred`, data);
   }
 
+  /**
+   * Check if there are third-party apps that
+   * subscribed for webhooks for a specific package
+   *
+   * @function _shouldSendWebhook
+   * @param {String} pkg - The name of the package
+   * @returns {Boolean} the computed value
+   */
+  _shouldSendWebhook(pkg) {
+    const webhooks = this.webhooks[pkg];
+    if (!webhooks || webhooks.length === 0) {
+      this.log.info(`No webhooks for pkg ${pkg}`);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Check if a build is in a queued status
+   *
+   * @function _isBuildQueued
+   * @param {Object} build - Build params
+   * @param {String} build.pkg - The package name
+   * @param {String} build.version - The package version
+   * @param {String} build.env - The build environnement
+   * @returns {String} the package name
+   */
+  async _isBuildQueued(build) {
+    // Fetch list of previous events for a certain build
+    const previousEvents = await this.models.Status.findAll(build);
+
+    // Ensure that last event is the 'queued' event from carpentd
+    const lastEvent = previousEvents[previousEvents.length - 1];
+    if (lastEvent && lastEvent.message === 'Builds Queued') return true;
+    return false;
+  }
+
+  /**
+   * Get package name from NSQ message
+   *
+   * @function _getPackageName
+   * @param {Object} data - Message from NSQ
+   * @returns {String} the package name
+   */
+  _getPackageName(data) {
+    return data.pkg || data.name;
+  }
+
+  /**
+   * Process and dispatch a webhook
+   *
+   * @function _dispatchWebhook
+   * @param {String} event - Webhook event type
+   * @param {Object} data - Message from NSQ
+   * @returns {Promise} to resolve
+   */
+  async _dispatchWebhook(event, data) {
+    const { version, env } = data;
+    const pkg = this._getPackageName(data);
+
+    if (!this._shouldSendWebhook(pkg)) return;
+
+    let body;
+
+    switch (event) {
+      case 'build_started':
+        if (!await this._isBuildQueued({ pkg, version, env })) return;
+        body = { event, pkg, version, env };
+        break;
+      default:
+        throw new Error(`'${event}' is not a valid webhook type`);
+    }
+
+    // Send webhook to subscribed third-parties
+    return this._sendWebhook(body);
+  }
+
+  /**
+   * Send webhook to third-party applications
+   *
+   * @function _sendWebhook
+   * @param {Object} body - Webhook body payload
+   * @returns {Promise} to resolve
+   */
+  async _sendWebhook(body) {
+    const webhooks = this.webhooks[body.pkg];
+    try {
+      const params = { body, method: 'POST', json: true };
+      await Promise.all(webhooks.map(uri => request({ uri, ...params })));
```

Comment:
we might want to wrap these with a retryme eventually but not urgent

Reply:
Sure, can do this separately.